sdk/clientcore/core/src/main/java/io/clientcore/core/util/union/Union.java

```diff
@@ -0,0 +1,283 @@
+     *
+     * @param consumer A consumer that will consume the value of the Union if it is of the expected type.
+     * @param clazz The expected type of the value.
+     * @return Returns true if the value was consumable by the consumer, and false if it was not.
+     * @param <T> The value type expected by the consumer.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> boolean tryConsume(Consumer<T> consumer, Class<T> clazz) {
+        if (isInstanceOfType(value, clazz)) {
+            consumer.accept(clazz.cast(value));
+            return true;
+        }
+
+        if (isPrimitiveTypeMatch(value, clazz)) {
+            consumer.accept((T) value);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * This method is used to consume the value of the Union if it is of the expected type.
+     *
+     * @param consumer A consumer that will consume the value of the Union if it is of the expected type.
+     * @param clazz The expected type of the value.
+     * @param genericTypes A var-args representation of generic types that are expected by the consumer, for example,
+     *                     <code>List&lt;String&gt;</code> would be represented as <pre>List.class, String.class</pre>.
+     * @return Returns true if the value was consumable by the consumer, and false if it was not.
+     * @param <T> The value type expected by the consumer.
+     */
+    public <T> boolean tryConsume(Consumer<T> consumer, Class<T> clazz, Class<?>... genericTypes) {
+        return tryConsume(consumer, new ParameterizedTypeImpl(clazz, genericTypes));
+    }
+
+    /**
+     * This method is used to consume the value of the Union if it is of the expected type.
+     *
+     * @param consumer A consumer that will consume the value of the Union if it is of the expected type.
+     * @param type The expected type of the value.
+     * @return Returns true if the value was consumable by the consumer, and false if it was not.
+     * @param <T> The value type expected by the consumer.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> boolean tryConsume(Consumer<T> consumer, ParameterizedType type) {
+        if (isInstanceOfType(value, type)) {
+            consumer.accept((T) value);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return value == null
+            ? "Union{types=" + types + ", value=null" + "}"
+            : "Union{types=" + types + ", type=" + currentType.getTypeName() + ", value=" + value + "}";
+    }
+
+    private boolean isInstanceOfType(Object value, Type type) {
+        if (type instanceof ParameterizedType) {
+            ParameterizedType pType = (ParameterizedType) type;
+            if (pType.getRawType()instanceof Class<?> rawType && rawType.isInstance(value)) {
+                Type[] actualTypeArguments = pType.getActualTypeArguments();
+                if (value instanceof Collection<?> collection) {
```

Comment:
If we're going to have limitations on what we're able to support would it be better to create explicit APIs with these limitations?

Reply:
offline discussion:
Will not introduce the explicit APIs but explain the restriction in more detail in javadoc.